code/01_spatial_overlay.R

```diff
@@ -0,0 +1,145 @@
+
+    clim_area_raster <- raster::rotate(clim_area_raster)
+
+  }
+  
+  # Check longitude ranges match (with a tolerance of 1 in case lon +- 179 vs. +-180)
+  poly_range <- c(extent(input_polygons)@xmin, extent(input_polygons)@xmax)
+  rast_range <- c(extent(clim_area_raster)@xmin, extent(clim_area_raster)@xmax)
+  
+  if(dplyr::near(poly_range[1], rast_range[1], tol=1.01) & dplyr::near(poly_range[2], rast_range[2], tol=1.01)){
+    
+    message(crayon::green('Longitude ranges match'))
+    
+  } else {
+    
+    stop(crayon::red('Raster and polygon longitude ranges do not match'))
+    
+  }
+  
+  # Match raster and polygon crs 
+  crs_raster <- crs(clim_area_raster)
+  polygons_reproj <- input_polygons %>% 
+    st_transform(crs = crs_raster)
+  
+  ## Geoweights (using data.table) 
+  ## -----------------------------------------------
+  message(crayon::yellow('Extracting raster polygon overlap'))
+    
+  geoweights <- rbindlist(exactextractr::exact_extract(clim_area_raster, polygons_reproj, progress = T, include_cell = T, include_xy = T), idcol = "poly_id")
+  geoweights[, ':=' (poly_id = polygons_reproj[[polygon_id]][poly_id], cell_area_km2 = value)] # Add the unique id for each polygon based on the input col name
+  geoweights <- geoweights[, .(gridNumber = cell, x, y, poly_id, w_geo = coverage_fraction * cell_area_km2)] # Geoweight = area km2 * coverage fraction 
+  geoweights[, w_geo := w_geo / sum(w_geo), by = poly_id] # Normalize by polygon
+
+  
+  message(crayon::yellow('Checking sum of weights within polygons'))
+  check_weights <- geoweights[, w_sum := sum(w_geo), by=poly_id]
+  
+  # Check that each polygon sums to 1, if not error w/polygon id 
+  for(i in nrow(check_weights)){
+    
+    if(check_weights$w_sum[i] != 1){
+      
+      stop(crayon::red('Area weights for polygon', check_weights$poly_id, 'do not sum to 1'))
+      
+    }
+    
+  }
+  # If it doesn't error out then all weight sums = 1
+  message(crayon::green('All weights sum to 1'))
+  
+  ## Save outputs 
+  ## -----------------------------------------------
+  
+  # Check if there is already a general geoweights folder
+  if(!dir.exists(here::here("data", "int", "geoweights"))){
+
+    # If no - create it
+    message(crayon::yellow('Creating data/int/geoweights/'))
+    dir.create(here::here("data", "int", "geoweights"), recursive=T)
+
+  }
+  
+  # File save name
+  polygon_input_name <- deparse(substitute(input_polygons))
+  save_name <- paste0(paste(polygon_input_name, data_source_norm, sep="_"), ".csv")
+  save_path <- file.path(here::here(), "data", "int", "geoweights")
+  
+  # Save message
+  message(crayon::yellow('Saving', save_name, 'to', save_path))
+  
+  # Save geoweights 
+  fwrite(geoweights[, ':=' (gridNumber = NULL, w_sum = NULL)], file = file.path(save_path, save_name))
+
+
+  ## Return weights 
+  ## -----------------------------------------------
+  return(geoweights)
+  
+}
+
+## Test function 
```

Comment:
I think I'd prefer a user script example that calls this function for a particular shapefile located on the server. Then a second user script example for the second stage as well. Or this could be in one script where you toggle at the top "do you need to do step 1?" since sometimes you'll only need to do step 2 and call the aggregate step.